### PR TITLE
(SLV-647) Ensure all puppet changes in post inst

### PR DESCRIPTION
**Pull request by johnduarte**
This commit updates the `30_final_puppet_run` pre-suite to ensure that
the puppet agent is run repeatedly until all changes are applied and the
exit code is `0`.

### CODE REVIEW

**Comment by johnduarte** [comment]
Test (results pending)
```
[john@gatling-runner jrd-gplt-01]$ git rev-parse HEAD
9843387ce7a919d6fcd0a7967603ee088df6ba0e
[john@gatling-runner jrd-gplt-01]$ env | grep 'BEAKER\|GATLING'
PUPPET_GATLING_SCALE_INCREMENT=0
BEAKER_TESTS=
PUPPET_GATLING_SCALE_ITERATIONS=1
PUPPET_GATLING_SCALE_BASE_INSTANCES=600
BEAKER_PE_DIR=http://enterprise.delivery.puppetlabs.net/2019.2/ci-ready
BEAKER_INSTALL_TYPE=pe
BEAKER_PE_VER=2019.2.0-rc10-23-gba5bc99
[john@gatling-runner jrd-gplt-01]$ bundle exec rake performance
...
```

File: setup/install_gatling/40_post_install/30_final_puppet_run.rb [hunk 1]
```diff
@@ -1,8 +1,17 @@
...
 
+MAX_AGENT_RUNS = 5
+
 test_name "final puppet run" do
-  step "run puppet on all nodes" do
+  step "Set server to master on hosts" do
     on hosts, "puppet config set --section agent server #{master.hostname}"
-    on hosts, "puppet agent -t", acceptable_exit_codes: [0, 2, 6]
+  end
+  step "Run puppet until all changes are applied" do
```
[the remark this answers is not in the document]
Comment (johnduarte) [reply setup/install_gatling/40_post_install/30_final_puppet_run.rb]: Thanks @billclaytor , I have put your suggested method in `perf_helper.rb`, added calls in `40_post_install/30_final_puppet_run.rb ` and  `50_tune/10_puppet_infrastructure_tune.rb`, and added spec tests for it.

File: spec/setup/helpers/perf_helper_spec.rb [hunk 2]
```diff
@@ -595,5 +595,22 @@ def step(_msg, &_block)
...
+    end
+    it "updates retry params based on given parameters" do
+      retries = 42
+      interval = 7
+      expected_retry_params = { max_retries: retries,
+                                retry_interval: interval,
+                                desired_exit_codes: [0] }
+      expect(subject).to receive(:puppet).with("agent", "-t")
+      expect(subject).to receive(:retry_on).with(hosts[0], nil, expected_retry_params)
+      expect(subject.run_agent_until_no_change(hosts, retries, interval)).to eq(hosts)
```
[the remark this answers is not in the document]
Comment (johnduarte) [reply spec/setup/helpers/perf_helper_spec.rb]: No we do not.  The expectation has been removed.

File: setup/install_gatling/50_tune/10_puppet_infrastructure_tune.rb [hunk 3]
```diff
@@ -45,12 +43,7 @@ def puppet_infrastructure_tune
     ].join(" ")
     on master, commit
 
-    # re-run the agent up to MAX_AGENT_RUNS attempts
-    retry_params = { max_retries: MAX_AGENT_RUNS,
-                     retry_interval: 2,
-                     desired_exit_codes: [0] }
-
-    retry_on(master, puppet("agent", "-t"), retry_params)
+    run_agent_until_no_change(master)
```
Comment (billclaytor) [inline setup/install_gatling/50_tune/10_puppet_infrastructure_tune.rb]: This actually fails:
```
NoMethodError: undefined method `each' for #<Unix::Host:0x00007f49ea763f78>
/home/centos/gplt/647/gatling-puppet-load-test/setup/helpers/perf_helper.rb:731
/home/centos/gplt/647/gatling-puppet-load-test/setup/install_gatling/50_tune/10_puppet_infrastructure_tune.rb:46
```

For some reason I thought beaker hosts could handle this; perhaps this needs to be called as:
```
run_agent_until_no_change([master])
```
?

Comment (johnduarte) [reply setup/install_gatling/50_tune/10_puppet_infrastructure_tune.rb]: :man_facepalming: sorry @billclaytor . This has been fixed.

**Pull request merged**